File: android/core/src/main/kotlin/com/paladex/ex/core/net/Http.kt
package com.paladex.ex.core.net

import java.io.IOException
import java.util.concurrent.TimeUnit
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

/** Non-2xx response from an upstream. Distinguishes "site said no" from "site said nothing". */
class HttpStatusException(val status: Int, val url: String) :
    IOException("HTTP $status for $url")

/**
 * Outbound HTTP with the manners a scraper owes the sites it reads: one request
 * per host at a time, a floor on the gap between them, a real User-Agent, and a
 * hard timeout so a hung socket cannot pin a coroutine open forever.
 */
class Http(
    private val minIntervalMillis: Long = 1_500,
    client: OkHttpClient? = null,
) {
    /** Chrome on macOS. Sites serve bot-flavoured HTML to anything obviously automated. */
    private val userAgent =
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 " +
            "(KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36"

    private val client: OkHttpClient = client ?: OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .callTimeout(20, TimeUnit.SECONDS)
        .followRedirects(true)
        .build()

    /** One mutex per host, so requests to a single host serialise. */
    private val hostLocks = mutableMapOf<String, Mutex>()
    private val hostLocksGuard = Mutex()
    private val lastRequestAt = mutableMapOf<String, Long>()

    private suspend fun lockFor(host: String): Mutex =
        hostLocksGuard.withLock { hostLocks.getOrPut(host) { Mutex() } }

    /**
     * Serialise work against [host], leaving at least [minIntervalMillis]
     * between consecutive requests.
     */
    private suspend fun <T> throttled(host: String, block: suspend () -> T): T =
        lockFor(host).withLock {
            val since = System.currentTimeMillis() - (lastRequestAt[host] ?: 0L)
            if (since < minIntervalMillis) delay(minIntervalMillis - since)
            try {
                block()
            } finally {
                lastRequestAt[host] = System.currentTimeMillis()
            }
        }

    private suspend fun execute(request: Request): String = withContext(Dispatchers.IO) {
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw HttpStatusException(response.code, request.url.toString())
            }
            response.body?.string().orEmpty()
        }
    }

    /**
     * Fetch with retry. Retries only on 429 and 5xx — other 4xx will not
     * improve by asking again.
     */
    private suspend fun withRetry(request: Request, retries: Int): String {
        var last: Exception? = null

        for (attempt in 0..retries) {
            // 1s, 2s, 4s — enough to clear a transient 429 without stalling the UI.
            if (attempt > 0) delay(1_000L shl (attempt - 1))

            try {
                return execute(request)
            } catch (e: HttpStatusException) {
                if (e.status != 429 && e.status < 500) throw e
                last = e
            } catch (e: IOException) {
                last = e
            }
        }

        throw last ?: IOException("Request to ${request.url} failed")
    }

    /**
     * @param immediate skips the per-host throttle. Use for first-party APIs,
     *   not for scraping.
     */
    suspend fun get(
        url: String,
        headers: Map<String, String> = emptyMap(),
        immediate: Boolean = false,
        retries: Int = 2,
    ): String {
        val request = Request.Builder()
            .url(url)
            .header("User-Agent", userAgent)
            .header("Accept-Language", "en-US,en;q=0.9")
            .apply { headers.forEach { (k, v) -> header(k, v) } }
            .build()

        return if (immediate) {
            withRetry(request, retries)
        } else {
            throttled(request.url.host) { withRetry(request, retries) }
        }
    }

    suspend fun postForm(
        url: String,
        body: String,
        headers: Map<String, String> = emptyMap(),
    ): String {
        val request = Request.Builder()
            .url(url)
            .post(body.toRequestBody("application/x-www-form-urlencoded".toMediaTypeOrNull()))
            .header("User-Agent", userAgent)
            .apply { headers.forEach { (k, v) -> header(k, v) } }
            .build()

        return withRetry(request, 1)
    }
}

File: src/lib/http.ts
import { env } from "./env";

/**
 * Outbound HTTP with the manners a scraper owes the sites it reads: one request
 * per host at a time, a floor on the gap between them, a real User-Agent, and a
 * hard timeout so a hung socket cannot pin a request handler open.
 */

const HTTP_TIMEOUT_MS = 15_000;

/** Chrome on macOS. Sites serve bot-flavoured HTML to anything obviously automated. */
const USER_AGENT =
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 " +
  "(KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36";

/** Tail of the in-flight chain per host, so requests to one host serialise. */
const hostQueues = new Map<string, Promise<void>>();

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Serialise work against `host`, leaving at least `minIntervalMs` between the
 * end of one task and the start of the next.
 */
function throttle<T>(host: string, task: () => Promise<T>, minIntervalMs: number): Promise<T> {
  const previous = hostQueues.get(host) ?? Promise.resolve();

  const run = previous.then(async () => {
    await sleep(minIntervalMs);
    return task();
  });

  // Keep the chain alive regardless of whether this task succeeded.
  hostQueues.set(
    host,
    run.then(
      () => undefined,
      () => undefined,
    ),
  );

  return run;
}

export interface FetchOptions {
  method?: "GET" | "POST";
  headers?: Record<string, string>;
  body?: string;
  /** Skip the per-host throttle. Use for first-party APIs, not for scraping. */
  immediate?: boolean;
  /** Retries on 429/5xx/network errors. */
  retries?: number;
}

export class HttpError extends Error {
  constructor(
    message: string,
    readonly status: number,
    readonly url: string,
  ) {
    super(message);
    this.name = "HttpError";
  }
}

async function once(url: string, options: FetchOptions): Promise<Response> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), HTTP_TIMEOUT_MS);

  try {
    return await fetch(url, {
      method: options.method ?? "GET",
      headers: {
        "User-Agent": USER_AGENT,
        "Accept-Language": "en-US,en;q=0.9",
        ...options.headers,
      },
      body: options.body,
      signal: controller.signal,
      redirect: "follow",
    });
  } finally {
    clearTimeout(timer);
  }
}

/**
 * Fetch with throttling and retry. Throws `HttpError` on a non-OK response so
 * providers can distinguish "site said no" from "site said nothing useful".
 */
export async function politeFetch(url: string, options: FetchOptions = {}): Promise<Response> {
  const host = new URL(url).host;
  const retries = options.retries ?? 2;

  const attempt = async (): Promise<Response> => {
    let lastError: unknown;

    for (let i = 0; i <= retries; i++) {
      if (i > 0) {
        // 1s, 2s, 4s — enough to clear a transient 429 without stalling the UI.
        await sleep(1000 * 2 ** (i - 1));
      }

      try {
        const response = await once(url, options);

        if (response.ok) return response;

        // 4xx other than 429 will not improve by asking again.
        if (response.status !== 429 && response.status < 500) {
          throw new HttpError(`${response.status} ${response.statusText}`, response.status, url);
        }

        lastError = new HttpError(`${response.status} ${response.statusText}`, response.status, url);
      } catch (error) {
        if (error instanceof HttpError && error.status !== 429 && error.status < 500) {
          throw error;
        }
        lastError = error;
      }
    }

    throw lastError instanceof Error ? lastError : new Error(`Request to ${url} failed`);
  };

  if (options.immediate) return attempt();
  return throttle(host, attempt, env.scrapeMinIntervalMs);
}

export async function fetchJson<T>(url: string, options: FetchOptions = {}): Promise<T> {
  const response = await politeFetch(url, {
    ...options,
    headers: { Accept: "application/json", ...options.headers },
  });
  return (await response.json()) as T;
}

export async function fetchHtml(url: string, options: FetchOptions = {}): Promise<string> {
  const response = await politeFetch(url, {
    ...options,
    headers: {
      Accept: "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
      ...options.headers,
    },
  });
  return response.text();
}
